server/integration-test/src/test/java/com/oceanbase/odc/service/userrecentbrowsing/UserRecentBrowsingServiceTest.java

```diff
@@ -0,0 +1,153 @@
+    }
+
+    @Test
+    public void getByProjectAndUserAndItem() {
+        UserRecentBrowsingEntity addedEntity =
+                userRecentBrowsingService.add(projectId, userId, itemType, itemId, new Date(browseMills));
+        UserRecentBrowsingEntity retrievedEntity =
+                userRecentBrowsingService.getByProjectAndUserAndItem(projectId, userId, itemType, itemId);
+
+        assertNotNull(retrievedEntity);
+        assertEquals(addedEntity.getId(), retrievedEntity.getId());
+        assertTimeEqualsInSeconds(browseMills, retrievedEntity.getBrowseTime());
+    }
+
+    @Test
+    public void listRecentBrowsingEntitiesWithLimit_Normal() {
+        UserRecentBrowsingEntity add1 = userRecentBrowsingService.add(projectId, userId, itemType, itemId,
+                new Date(browseMills));
+        UserRecentBrowsingEntity add2 = userRecentBrowsingService.add(projectId, userId, BrowseItemType.GIT_REPO,
+                itemId + 1,
+                new Date(browseMills + 1000));
+        UserRecentBrowsingEntity add3 = userRecentBrowsingService.add(projectId, userId, itemType,
+                itemId + 2,
+                new Date(browseMills - 3000));
+
+        List<UserRecentBrowsingEntity> recentEntities = userRecentBrowsingService
+                .listRecentBrowsingEntitiesWithLimit(projectId, userId,
+                        Arrays.asList(BrowseItemType.GIT_REPO, BrowseItemType.WORKSHEET), 100);
+
+        assertEquals(3, recentEntities.size());
+        assertEquals(add2.getId(), recentEntities.get(0).getId());
+        assertEquals(add1.getId(), recentEntities.get(1).getId());
+        assertEquals(add3.getId(), recentEntities.get(2).getId());
+    }
+
+    @Test
+    public void listRecentBrowsingEntitiesWithLimit_ExceedLimit() {
+        for (int i = 0; i < 100; i++) {
+            userRecentBrowsingService.add(projectId, userId, itemType, itemId + i,
+                    new Date(browseMills + i * 1000));
+            userRecentBrowsingService.add(projectId, userId, BrowseItemType.GIT_REPO,
+                    itemId + 100 + i,
+                    new Date(browseMills + i * 1000));
+        }
+
+        List<UserRecentBrowsingEntity> recentEntities = userRecentBrowsingService
+                .listRecentBrowsingEntitiesWithLimit(projectId, userId,
+                        Arrays.asList(BrowseItemType.GIT_REPO, BrowseItemType.WORKSHEET), 100);
+
+        assertEquals(100, recentEntities.size());
+        assertTimeEqualsInSeconds(browseMills + 50 * 1000, recentEntities.get(98).getBrowseTime());
+        assertTimeEqualsInSeconds(browseMills + 50 * 1000, recentEntities.get(99).getBrowseTime());
+
+        assertEquals(BrowseItemType.GIT_REPO.name(), recentEntities.get(98).getItemType());
+        assertEquals(BrowseItemType.WORKSHEET.name(), recentEntities.get(99).getItemType());
+    }
+
+    private void assertTimeEqualsInSeconds(Long ts, Date time) {
```

Comment:
unit testing should contains no logic

Reply:
Due to the fact that the precision of database timestamps is in seconds, this is done to compare the precision in seconds